### ISSUE TITLE
Issue with Required Validation

### ISSUE THREAD

**Issue by natebrunette** [justinrainbow/json-schema#254]
In `UndefinedConstraint::validatedCommonProperties()` if `$value` is not an object, it will assume that the value is valid.  I've run into multiple scenarios where the value is erroneously an array and the schema appears to be valid, even though it actually isn't.

**Comment by jojo1981** [comment]
@natebrunette can you add an example schema file and an example test data file? So I can try to reproduce this issue.

**Comment by natebrunette** [comment]
I'll work on extracting my schema, I've also created a pull request to address this issue.

**Comment by natebrunette** [comment]
Here's the definition I'm trying to validate against

```
"Foo": {
    "properties": {
        "data": {
            "type": "array",
            "items": {
                "properties": {
                    "id": {
                        "type": "integer"
                    },
                    "title": {
                        "type": "string"
                    }
                },
                "required": [
                    "id",
                    "title"
                ]
            }
        }
    },
    "required": [
        "data"
    ]
}
```

And here's an example json

```
{
    "data": [
        [
            "Error"
        ],
        [
            "Error"
        ]
    ]
}
```

**Comment by natebrunette** [comment]
There's still an issue with the above pull request trying to validate non-required parent fields.

**Comment by jojo1981** [comment]
@natebrunette I've tested the passed schema and data set with the newest version (2.0.0) of this library and must conclude that this is indeed a bug.

Thanks for the pull request I will join the review!

**Comment by jojo1981** [comment]
@natebrunette After reviewing the code and test some things I must conclude this is **NOT** a bug. The problem here is that the root level is not a schema. "Foo" here is a subschema but there is no reference to it. So it will be ignored. We end up with an empty schema. Validating an empty schema is useless but will always return true.

Tested by converting the schema into:
    
    {
      "properties": {
        "data": {
          "type": "array",
          "items": {
            "type": "object",
            "properties": {
              "id": {
                "type": "integer"
              },
              "title": {
                "type": "string"
              }
            },
            "required": [
              "id",
              "title"
            ]
          }
        }
      },
      "required": [
        "data"
      ]
    }
    
and test with data:
    
    {
      "data": [
        [
          "Error"
        ],
        [
          "Error"
        ]
      ]
    }
    
Will result in:
    
    [
        {
            "property": "data[0]",
            "message": "Array value found, but an object is required",
            "constraint": "type"
        },
        {
            "property": "data[1]",
            "message": "Array value found, but an object is required",
            "constraint": "type"
        }
    ]
    
I think this is just the way it should be!

**Comment by natebrunette** [comment]
Ahh, this makes sense.  I didn't realize I had to specify the type as an object.  Thanks for your help.

**Comment by jojo1981** [comment]
@natebrunette Thanks for closing this issue. I think the corresponding PR can also be closed? If so would you please close that PR (#255). Thanks in advance!
